src/pudl/output/ferc1.py

```diff
@@ -871,3 +873,287 @@ def add_mean_cap_additions(steam_df):
         )
     )
     return df
+
+
+#########
+# Explode
+#########
+
+
+def explode_tables(
+    tables_to_explode: list[str],
+    tables: dict[str : pd.DataFrame],
+    top_table: str,
+    clean_xbrl_metadata_json: dict,
+) -> pd.DataFrame:
+    """Explode a set of nest tables."""
+    # but really just feed dagster the right list of assets
+    # tables = {tbl: defs.load_asset_value(AssetKey(tbl)) for tbl in tables_to_explode}
```

Comment:
If I understand correctly, you want to create an asset for each of these sets of tables. If so, that sounds kind of like the `ferc1_transform_asset_factory` and might look something like:

```py
from dagster import AssetIn, AssetsDefinition, asset


def explode_table_asset_factory(
    asset_name: str,
    input_tables: list[str]
) -> AssetsDefinition:
    ins: = {"clean_xbrl_metadata_json": AssetIn("clean_xbrl_metadata_json")}
    ins |= {table_name: AssetIn(table_name) for table_name in input_tables}

    @asset(name=asset_name, ins=ins)
    def explode_tables(
         clean_xbrl_metadata_json: dict,
         **tables: dict[str : pd.DataFrame],
    ) -> pd.DataFrame:
        tables_to_explode = list(tables.keys())
        # Do the things here


def create_explode_table_assets() -> list[AssetsDefinition]:
    income_statement_tables = [ "depreciation_amortization_summary_ferc1", "income_statement_ferc1", "electric_operating_expenses_ferc1", "electric_operating_revenues_ferc1" ]
    balance_sheet_asset_tables = [ "balance_sheet_assets_ferc1" "utility_plant_summary_ferc1", "plant_in_service_ferc1", "electric_plant_depreciation_changes_ferc1", ]
    balance_sheet_liability_tables = [ "balance_sheet_liabilities_ferc1", "retained_earnings_ferc1" ]

    assets = [
        explode_table_asset_factory("income_statement_exploded", income_statement_tables),
        explode_table_asset_factory("balance_sheet_exploded", balance_sheet_asset_tables),
        explode_table_asset_factory("balance_sheet_liability_exploded", balance_sheet_liability_tables),
    ]
    return assets


explode_assets = create_explode_table_assets()
```

Reply:
ty @zschira !!  I'm getting an error which I believe means it isn't able to find a particular asset from these lists of ferc tables as its compiling `ins`

```
2023-06-06 12:18:13 -0400 - dagster.code_server - INFO - Started Dagster code server for module pudl.ferc_to_sqlite in process 21844
/Users/christinagosnell/mambaforge/envs/pudl-dev/lib/python3.11/site-packages/dagster/_core/workspace/context.py:594: UserWarning: Error loading repository location pudl.ferc_to_sqlite:AttributeError: 'dict' object has no attribute 'key'

Stack Trace:
  File "/Users/christinagosnell/mambaforge/envs/pudl-dev/lib/python3.11/site-packages/dagster/_grpc/server.py", line 278, in __init__
    self._loaded_repositories: Optional[LoadedRepositories] = LoadedRepositories(
                                                              ^^^^^^^^^^^^^^^^^^^
  File "/Users/christinagosnell/mambaforge/envs/pudl-dev/lib/python3.11/site-packages/dagster/_grpc/server.py", line 122, in __init__
    loadable_targets = get_loadable_targets(
                       ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/christinagosnell/mambaforge/envs/pudl-dev/lib/python3.11/site-packages/dagster/_grpc/utils.py", line 47, in get_loadable_targets
    else loadable_targets_from_python_module(module_name, working_directory)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/christinagosnell/mambaforge/envs/pudl-dev/lib/python3.11/site-packages/dagster/_core/workspace/autodiscovery.py", line 35, in loadable_targets_from_python_module
    module = load_python_module(
             ^^^^^^^^^^^^^^^^^^^
  File "/Users/christinagosnell/mambaforge/envs/pudl-dev/lib/python3.11/site-packages/dagster/_core/code_pointer.py", line 135, in load_python_module
    return importlib.import_module(module_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/christinagosnell/mambaforge/envs/pudl-dev/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1128, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/Users/christinagosnell/code/pudl/src/pudl/__init__.py", line 5, in <module>
    from . import (  # noqa: F401
  File "/Users/christinagosnell/code/pudl/src/pudl/analysis/__init__.py", line 8, in <module>
    from . import (  # noqa: F401
  File "/Users/christinagosnell/code/pudl/src/pudl/analysis/state_demand.py", line 37, in <module>
    import pudl.output.pudltabl
  File "/Users/christinagosnell/code/pudl/src/pudl/output/__init__.py", line 19, in <module>
    from . import ferc1  # noqa: F401
    ^^^^^^^^^^^^^^^^^^^
  File "/Users/christinagosnell/code/pudl/src/pudl/output/ferc1.py", line 966, in <module>
    exploded_ferc1_assets = create_explode_table_assets()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/christinagosnell/code/pudl/src/pudl/output/ferc1.py", line 953, in create_explode_table_assets
    explode_table_asset_factory(
  File "/Users/christinagosnell/code/pudl/src/pudl/output/ferc1.py", line 893, in explode_table_asset_factory
    @asset(name=asset_name, ins=ins)  # , io_manager_key=io_manager_key)
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/christinagosnell/mambaforge/envs/pudl-dev/lib/python3.11/site-packages/dagster/_core/definitions/decorators/asset_decorator.py", line 227, in inner
    return create_asset()(fn)
           ^^^^^^^^^^^^^^^^^^
  File "/Users/christinagosnell/mambaforge/envs/pudl-dev/lib/python3.11/site-packages/dagster/_core/definitions/decorators/asset_decorator.py", line 293, in __call__
    asset_ins = build_asset_ins(fn, self.ins or {}, self.non_argument_deps)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/christinagosnell/mambaforge/envs/pudl-dev/lib/python3.11/site-packages/dagster/_core/definitions/decorators/asset_decorator.py", line 654, in build_asset_ins
    asset_key = asset_ins[input_name].key
                ^^^^^^^^^^^^^^^^^^^^^^^^^

  warnings.warn(
```

Reply:
also this may be a basic question but how does `explode_tables` know to decompose `ins` as `clean_xbrl_metadata_json` and the rest of the tables? is it just grabbing `clean_xbrl_metadata_json` by name and assuming everything else is a table?

Reply:
i figured it out it was a missing comma 😎

Reply:
Stupid commas!

> also this may be a basic question but how does explode_tables know to decompose ins as clean_xbrl_metadata_json and the rest of the tables? is it just grabbing clean_xbrl_metadata_json by name and assuming everything else is a table?

I think that's basically right. I'm not sure exactly how Dagster handles the `ins` under the hood, but in general you can do something like:

```py
def func(a, **kwargs):
    # Do stuff

func(**{"a": 0, "other_args": 1})
```
and python will handle mapping `a` from the dict to the arg `a`.